---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# KRIS <img src="man/figures/kris_logo.png" align="right" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/kridsadakorn/kris/workflows/R-CMD-check/badge.svg)](https://github.com/kridsadakorn/kris/actions)
[![CRAN status](https://www.r-pkg.org/badges/version/KRIS)](https://CRAN.R-project.org/package=KRIS)
[![codecov](https://codecov.io/gh/kridsadakorn/kris/branch/master/graph/badge.svg?token=QO20Q25CL6)](https://codecov.io/gh/kridsadakorn/kris)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Lifecycle: maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![DOI](https://zenodo.org/badge/333291793.svg)](https://zenodo.org/badge/latestdoi/333291793)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fkridsadakorn%2Fkris.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fkridsadakorn%2Fkris?ref=badge_shield)
<!-- badges: end -->

```KRIS``` (**K**een and **R**eliable **I**nterface **S**ubroutines for 
bioinformatic analysis) is the package providing useful functions which are 
needed for bioinformatic analysis such as calculating linear principal 
components from numeric data and Single-nucleotide polymorphism (SNP) dataset, 
calculating fixation index (Fst) using Hudson method, creating scatter plots in 
3 views, handling with PLINK binary file format, detecting rough structures and 
outliers using unsupervised clustering, and calculating matrix multiplication 
in the faster way for big data..

## Installation

You can install the released version of KRIS from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("KRIS")
```

Alternatively, you can install the dev version of KRIS from [Gitlab](https://github.com/kridsadakorn/kris) with
``` r
install.packages("remotes")
remotes::install_github("kridsadakorn/kris", dependencies = TRUE)
```

## Document

You can see the reference manual from: [https://www.biostatgen.org/kris/](https://www.biostatgen.org/kris/)


## Example

This is a basic example which shows you how to use the package:

```{r example}
library(KRIS)

data(example_SNP)
```

```example_SNP``` consists of the simulated SNP data ```simsnp``` and the sample labels ```sample_labels```.

```{r example_SNP}
summary(simsnp)
summary(sample_labels)
```

To calcualte the first three linear principal components from SNP data:

```{r linear_PCA}
PCs <- cal.pc.linear(simsnp$snp, no.pc = 3)
```

You can visualize 3 dimensions for the first three principal components as:

```{r plotpc, echo=FALSE}
plot3views( PCs$PC, sample_labels)
```

## License
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fkridsadakorn%2Fkris.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fkridsadakorn%2Fkris?ref=badge_large)

## About

* Prof. Kristel Van Steen, visit  <a href="http://bio3.giga.ulg.ac.be/" border=0 style="border:0; text-decoration:none; outline:none"><img width="40px" src="man/figures/bio3_logo.png" align="center" /></a><br />
* Kridsadakorn Chaichoompu, visit <a href="https://www.biostatgen.org/" border=0 style="border:0; text-decoration:none; outline:none"><img width="110px" src="man/figures/biostatgen_logo.png" align="center" /></a><br />
